### PR TITLE
Fix a scatter eager_subs issue

### PR DESCRIPTION
**Pull request by fehiepsi**
Pair-coded with @eb8680 @fritzo @ordabayevy 

This bug is detected while running a test for infer_discrete with NumPyro scan

### CODE REVIEW

File: funsor/terms.py [hunk 1]
```diff
@@ -1218,6 +1218,17 @@ def _alpha_convert(self, alpha_subs):
         reduced_vars = frozenset(alpha_subs.get(var.name, var) for var in reduced_vars)
         return op, subs, source, reduced_vars
 
+    def eager_subs(self, subs):
+        subs = OrderedDict(subs)
+        new_subs = []
+        for name, sub in self.subs:
+            if name in subs:
+                assert isinstance(subs[name], Variable)
```
Comment (fehiepsi) [inline funsor/terms.py:1226]: Remove this assertation seem to resolve the Slice substitution issue

(there is another issue happens in `test_scan_hmm` that `@_get_support_value.register(funsor.cnf.Contraction)` couldn't extract the `delta_terms` - but I guess that issue is not related to this)

Comment (eb8680) [reply funsor/terms.py:1226]: Hmm, I guess what's happening is that there can be `bound` variables of `Scatter` that share names with the `fresh` variables, and we need to implement a better `_alpha_convert` method that properly handles this case. Rather than just removing the assertion, does changing this to
```py
if name in subs and isinstance(subs[name], Variable):
    new_subs.append((subs[name].name, sub))
```
fix the downstream error with `Slice`?  It's still not correct in general, but it might be enough to unblock you.

Comment (fehiepsi) [reply funsor/terms.py:1226]: Yes, it does resolving the Slice issue (but the test still fails by the `_get_support_value` issue mentioned above).

**Review by eb8680 — APPROVED** [review]
